### PR TITLE
2648 add `RandCoarseDropout` transform

### LINKED ISSUE #2185: Cutout data augmentation
**Is your feature request related to a problem? Please describe.**
"CoarseDropout of the square regions in the image"

2D version is included in albumentations:
https://github.com/albumentations-team/albumentations/blob/95a048d342cbe748e5acf15cb1a28611a6303885/albumentations/augmentations/transforms.py#L608-L609
would be great to have an ND version in this repo.

see also: https://arxiv.org/abs/1708.04552

### PR DESCRIPTION
**Pull request by Nic-Ma**
part of #2648 .
fixes #2185

### Description
This PR implemented the `RandCoarseDropout` and `RandCoarseDropoutd` transforms.

### Status
**Work in progress**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by wyli** [comment]
Will this address https://github.com/Project-MONAI/MONAI/issues/2185?

**Comment by Nic-Ma** [comment]
> Will this address #2185?

Yes, I think so.
If you feel any cases missing, let me add the support.

Thanks.

File: monai/transforms/intensity/array.py [hunk 1]
```diff
@@ -1603,3 +1606,61 @@ def _to_numpy(self, img: Union[np.ndarray, torch.Tensor]) -> Tuple[np.ndarray, t
...
+        self.hole_coords: List = []
+
+    def randomize(self, img_size: Sequence[int]) -> None:
+        super().randomize(None)
+        size = fall_back_tuple(self.spatial_size, img_size)
+        self.hole_coords = []  # clear previously computed coords
+        for _ in range(self.holes if self.max_holes is None else self.R.randint(self.holes, self.max_holes + 1)):
+            if self.max_spatial_size is not None:
+                max_size = fall_back_tuple(self.max_spatial_size, img_size)
+                size = tuple([self.R.randint(low=size[i], high=max_size[i] + 1) for i in range(len(img_size))])
```
Comment (wyli) [inline monai/transforms/intensity/array.py:1656]: the `[` and `]` are not needed

**Review by wyli — APPROVED** [review]

File: monai/transforms/intensity/array.py [hunk 2]
```diff
@@ -1603,3 +1606,61 @@ def _to_numpy(self, img: Union[np.ndarray, torch.Tensor]) -> Tuple[np.ndarray, t
             return img.cpu().detach().numpy(), img.device
         else:
             return img, torch.device("cpu")
+
+
+class RandCoarseDropout(RandomizableTransform):
+    """
+    Randomly coarse dropout regions in the image, then fill in the "holes" with specified value.
```
Comment (ericspod) [inline monai/transforms/intensity/array.py:1613]: -> "Randomly coarse dropout rectangular regions in the image, then fill in the "holes" with a specified value."

Should emphasize that it's rectangular regions being filled in.

File: monai/transforms/intensity/array.py [hunk 3]
```diff
@@ -1603,3 +1606,61 @@ def _to_numpy(self, img: Union[np.ndarray, torch.Tensor]) -> Tuple[np.ndarray, t
...
+    Refer to: https://arxiv.org/abs/1708.04552 and:
+    https://albumentations.ai/docs/api_reference/augmentations/transforms/
+    #albumentations.augmentations.transforms.CoarseDropout.
+
+    Args:
+        holes: number of regions to dropout, if `max_holes` is not None, use this arg as the minimum number to
+            randomly select the expected number of regions.
+        spatial_size: spatial size of the regions to dropput, if `max_spatial_size` is not None, use this arg
```
Comment (ericspod) [inline monai/transforms/intensity/array.py:1621]: "dropput" -> "dropout"

File: monai/transforms/intensity/array.py [hunk 4]
```diff
@@ -1603,3 +1606,61 @@ def _to_numpy(self, img: Union[np.ndarray, torch.Tensor]) -> Tuple[np.ndarray, t
...
+        self.fill_value = fill_value
+        self.max_holes = max_holes
+        self.max_spatial_size = max_spatial_size
+        self.hole_coords: List = []
+
+    def randomize(self, img_size: Sequence[int]) -> None:
+        super().randomize(None)
+        size = fall_back_tuple(self.spatial_size, img_size)
+        self.hole_coords = []  # clear previously computed coords
+        for _ in range(self.holes if self.max_holes is None else self.R.randint(self.holes, self.max_holes + 1)):
```
Comment (ericspod) [inline monai/transforms/intensity/array.py:1653]: I would break this into two lines to aid readability:

```python
num_holes = self.holes if self.max_holes is None else self.R.randint(self.holes, self.max_holes + 1)
for _ in range(num_holes):
```

File: monai/transforms/intensity/dictionary.py [hunk 5]
```diff
@@ -1324,6 +1330,71 @@ def _to_numpy(self, d: Union[torch.Tensor, np.ndarray]) -> np.ndarray:
...
+        self.fill_value = fill_value
+        self.max_holes = max_holes
+        self.max_spatial_size = max_spatial_size
+        self.hole_coords: List = []
+
+    def randomize(self, img_size: Sequence[int]) -> None:
+        super().randomize(None)
+        size = fall_back_tuple(self.spatial_size, img_size)
+        self.hole_coords = []  # clear previously computed coords
+        for _ in range(self.holes if self.max_holes is None else self.R.randint(self.holes, self.max_holes + 1)):
```
Comment (ericspod) [inline monai/transforms/intensity/dictionary.py:1380]: Same change here for readability.

File: monai/transforms/intensity/array.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/transforms/intensity/array.py:1656]: sure, updated in the latest commit.
Thanks.

File: monai/transforms/intensity/array.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/transforms/intensity/array.py:1613]: Sure, updated in the latest commit.
Thanks.

File: monai/transforms/intensity/array.py (hunk 3 above)
Comment (Nic-Ma) [reply monai/transforms/intensity/array.py:1621]: Thanks for pointing out the typo, updated in the latest commit.

File: monai/transforms/intensity/dictionary.py (hunk 5 above)
Comment (Nic-Ma) [reply monai/transforms/intensity/dictionary.py]: Sure, updated in the latest commit.
Thanks.

File: monai/transforms/intensity/array.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/transforms/intensity/array.py]: Sure, updated in the latest commit.
Thanks.
